fv3config/config/derive.py

```diff
@@ -42,33 +42,115 @@ def get_run_duration(config):
         if coupler_res_filename is not None:
-            current_date = _get_current_date_from_coupler_res(coupler_res_filename)
+            date = coupler_res_date_getter(coupler_res_filename)
         else:
-            current_date = config["namelist"]["coupler_nml"].get(
+            date = config["namelist"]["coupler_nml"].get(
                 "current_date", [0, 0, 0, 0, 0, 0]
             )
-    return current_date
+    return date
+
+
+def get_current_date(config):
+    """Return current_date from configuration dictionary. This function may read from
+    the remote initial_conditions path in the given configuration dictionary.
+
+    Args:
+        config (dict): a configuration dictionary
+
+    Returns:
+        list: current_date as list of ints [year, month, day, hour, min, sec]
+    """
+    return _get_date(config, _get_current_date_from_coupler_res)
+
+
+def get_diag_table_base_date(config):
+    """Return base_date from configuration dictionary. This function may read from
+    the remote initial_conditions path in the given configuration dictionary.
+
+    In the event that force_date_from_namelist is specified in the config, this
+    function returns the current_date or a null date if that does not exist.  If the
+    date is not set to be forced from the namelist, this function first looks for the
+    existence of a coupler.res file, and if it exists, returns the initialization_date
+    written within it. If a coupler.res files does not exist, the current_date
+    specified in the namelist or a null date is used.  To FV3GFS, a null date and
+    the current date are equivalent.
+
+    Returning the initialization_date instead of the current_date if a coupler.res
+    file exists is important for obtaining reproducible restarts in segmented runs.
+    Another advantage is that it harmonizes the units used to encode times in
+    fotran-generated diagnostics files across segments. See more discussion in
+    GH 147.
+
+    Args:
+        config (dict): a configuration dictionary
+
+    Returns:
+        list: base_date as list of ints [year, month, day, hour, min, sec]
+    """
+    return _get_date(config, _get_initialization_date_from_coupler_res)
+
+
+def _parse_date_from_line(line, coupler_res_filename):
+    """Parses a date from a line in a coupler.res file"""
+    date = [int(d) for d in re.findall(r"\d+", line)]
+    if len(date) != 6:
+        raise ConfigError(
+            f"{coupler_res_filename} does not have a valid date in the given line "
+            f"(line must contain six integers)"
+        )
+    return date
+
+
+def _read_coupler_res_dates(coupler_res_filename):
+    """Read the dates contained in a coupler.res file
+    
+    Args:
+        coupler_res_filename (str): a coupler.res filename
+
+    Returns:
+        initialization_date (list): date as list of ints [year, month, day, hour, min, sec]
+        current_date (list): date as list of ints [year, month, day, hour, min, sec]
```

Comment:
Lot's of redundant text here that could be expressed in types.

Reply:
e.g. by returning a `datetime.datetime` object.

Reply:
this is an optional change.

Reply:
Thanks -- I will hold off on this for now.  While I'm not sure if it is possible to run FV3GFS with a calendar other than "julian" (or whether we care to support that) the `coupler.res` file does include calendar type information, which we might make use of in constructing dates:
```
    2        (Calendar: no_calendar=0, thirty_day_months=1, julian=2, gregorian=3, noleap=4)
  2016     8     1     0     0     0        Model start time:   year, month, day, hour, minute, second
  2016     8     3     0     0     0        Current model time: year, month, day, hour, minute, second
```
For the time being I think sticking with the calendar-agnostic list that was used previously is fine.

Reply:
Thanks -- I will hold off on this for now.  While I'm not sure if it is possible to run FV3GFS with a calendar other than "julian" (or whether we care to support that) the `coupler.res` file does include calendar type information, which we might make use of in constructing dates:
```
    2        (Calendar: no_calendar=0, thirty_day_months=1, julian=2, gregorian=3, noleap=4)
  2016     8     1     0     0     0        Model start time:   year, month, day, hour, minute, second
  2016     8     3     0     0     0        Current model time: year, month, day, hour, minute, second
```
For the time being I think sticking with the calendar-agnostic list that was used previously is fine.